sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1005,7 +1006,7 @@ class DataFrame private[sql](
    * @since 1.3.0
    */
   def unionAll(other: DataFrame): DataFrame = withPlan {
-    Union(logicalPlan, other.logicalPlan)
+    CombineUnions(Union(logicalPlan, other.logicalPlan))
```

Comment:
Michael suggested this actually.

We should add some inline comment explaining that this breaks caching, but it's usually ok because it addresses a very specific use case: using union to union many files or partitions.